### ISSUE TITLE
Forecasting and time series methods

### ISSUE THREAD

**Issue by alexhallam** [mlpack/mlpack#2668]
### What is the desired addition or change?

It would be nice to have some time series methods. Like 

1. naive
2. seasonal naive
3. seasonal trend loess decomposition.
4. Holt Winters
6. Exponential smoothing
7. Arima

Since many time series come in groups it may be useful to think about how the data should be organized to take in many time series and account for seasonality. Also probabilistic forecasting is becoming more popular.  When possible each time series forecast should generate probabilistic forecasts.

### What is the motivation for this feature?

Forecasting is very common. 

### If applicable, describe how this feature would be implemented.

Each method as different needs.

**Comment by alexhallam** [comment]
I have experience in forecasting. If there is interest in adding time series and other forecasting methods to mlpack I would be happy to contribute.

**Comment by jeffin143** [comment]
@alexhallam , Yes we would be very happy to have those integrated in our library, please go ahead if you have the bandwidth to implement some of those also welcome to the community :)

**Comment by Abilityguy** [comment]
This is a great idea. 
I have previously worked on time series methods for a project.
I could help out with implementing a few methods if needed.

**Comment by alexhallam** [comment]
I am slowly dipping my toe in the water here with the simple mean forecast. I have a couple of questions. First let me explain what I am doing and what I want to do.

Given a users time series data

```
ds                        y
2020-01-01        1
2020-01-02        2
2020-01-02        1
```

the function asks for the desired percentiles, forecast horizon, number of bootstraps to get the residual, etc. The output looks like this

```
    1.0000    0.1000    0.4093
    1.0000    0.2000    0.7177
    1.0000    0.3000    0.7493
    1.0000    0.4000    0.9420
    1.0000    0.5000    1.2806
    1.0000    0.6000    1.5256
    1.0000    0.7000    1.7067
    1.0000    0.8000    2.2490
    1.0000    0.9000    2.6614
    2.0000    0.1000    0.2359
    2.0000    0.2000    0.6994
    2.0000    0.3000    0.7346
    2.0000    0.4000    0.9355
    2.0000    0.5000    1.2806
    2.0000    0.6000    1.5218
    2.0000    0.7000    1.6399
    2.0000    0.8000    2.2490
    2.0000    0.9000    2.5052
```

where the first column is the time index, the second column is the percentile of interest and the third column is the forecast.

I have the following question:

1. Bootstraps - I would rather have them be an option as oppose to a requirement. The default would be a t-distribution. Can I import the boot package header file to use the t-distribution?

2. Box-Cox Transformation - Does anyone know of a library that implements box-cox? This would improve the quality of the forecast in some cases. 

3. Dates - I don't like the index column I am showing above.  I would rather have dates than a time index. I found that Boost has date functionality, but it requires linking. If someone has a better date solution let me know.

**Comment by rcurtin** [comment]
Hey @alexhallam, I'm really sorry for the slow response on this one!

1. I think you mean boost not boot?  Anyway we're actually trying to remove our boost dependencies, because it can make compilation much, much slower.  See #2440 for more information.  Sorry if that makes things a little bit inconvenient!  Maybe we can find some other solution?

2. I think the box-cox transformation is pretty simple if I remember right.  Maybe we can just implement it here?

3. Well, we actually have two issues here.  The first is that C++ doesn't really provide a good date structure.  The second is that Armadillo, the matrix library we use, can only represent primitive numeric types inside of the matrix.  We do have some support for mapping dates to integers just like you did in your example, but it would expect the dates to be strings, and we would lose the timestamp quality of it.  I'm really not sure what the best solution here is.  At the lowest level, though, we should probably implement any time-series methods to take in time (represented as an arbitrary unit) as a Float64.  Or, we would have to allow the user to pass not just a data matrix, but also the timestamps corresponding with each point in the data matrix.  I'm really not sure here but it probably warrants some thought and discussion.  What do you think?  Sorry for the hidden complexity! :)

**Comment by alexhallam** [comment]
1. Yes, sorry,  I meant boost. I understand the complexity that this adds. For me this is probably the biggest blocker. I am a big fan of probabilistic forecasting which requires distributions. 

2. Box-Cox is not hard at all, auto-box-cox would require ensmallen for some optimizations. I was mostly curious if the goal is to have a "transformations" module within mlpack so box-cox could be used generally as oppose to just within my function.

3. Allowing the user to pass a data matrix and a date matrix is the best solution I can think of. I have worked with float representations of dates before and it is not ergonomic. Building a good date structure in C++ is outside of my skill set.

So in summary, I can add a box-cox transformation to mlpack, just let me know where to put it. In fact, maybe it is worth opening another issue on transformations in general. I could do some work in that space. When it comes to forecasting methods I am going to hold off.

Also, @Abilityguy feel free to run with any forecasting method you want. It is very viable to use just a data matrix and float representations of dates to generate point forecasts. Even that would be a nice feature in mlpack. This is just something I am not too interested in.

**Comment by Abilityguy** [comment]
Thanks @alexhallam , I haven't been able to pick this up until now, I will try to get a working PR of something soon.

**Comment by heisenbuug** [comment]
Hello @alexhallam @Abilityguy, I really liked the idea of adding forecasting methods, I went through this [page](https://pawarbi.github.io/blog/forecasting/r/python/rpy2/altair/fbprophet/ensemble_forecast/uncertainty/simulation/2020/04/21/timeseries-part2.html#:~:text=Seasonal%20naive%20method%20uses%20the,but%20helps%20create%20a%20baseline.), it has implementations and explains the basic stuff, I am thinking of taking up naive and seasonal naive...

**Comment by jeffin143** [comment]
> we would lose the timestamp quality of it 

May be we can write a function to convert date to unix time stamps and then do normal processing : https://www.epochconverter.com/

**Comment by alexhallam** [comment]
@jeffin143 I think that date processing is a great idea. I think we can be opinionated with the date string "YYYY-MM-DD HH:mm:ss" and throw an error if dates are not in that format. With that we can work in unix time then convert back to strings for the output. 

@heisenbuug I like that page. It has nice clear code. For something on the technical side there is this book https://otexts.com/fpp2/simple-methods.html. Anything by Rob Hyndman is quality forecasting work https://robjhyndman.com/.

**Comment by heisenbuug** [comment]
Hey @jeffin143 @Abilityguy @alexhallam So, I implemented a class for time conversion between `Unix Epoch Format` to `Human Readable Format`. For human-readable, I chose `struct tm` type. You can see the implementation [here](https://github.com/heisenbuug/Forecast-Utils_For_Mlpack/blob/main/TimeConversion.cpp)

- Converted Unix epoch -> Human readable
- Will start to work on Human readable -> Unix Epoch

**Comment by kushagrabhushan** [comment]
Hi. I'm very new to open source. Can someone tell me how to get started? I'd really like to contribute!

**Comment by rcurtin** [comment]
@alexhallam @jeffin143 yeah, I think UNIX timestamps could work to start with.  It would be great to eventually have support for more than just UNIX timestamps, but, perhaps that can be part of some other effort.  (For instance, a "data frame" built on top of Armadillo could be built to support arbitrary types, like dates or strings, and then "under the hood" when passing to mlpack algorithms, use whatever representation is best.  But, like I said, maybe that is very far off!  It would be a lot of work.)

**Comment by RishabhGarg108** [comment]
@alexhallam @rcurtin , I would like to work on "naive model". Shall I go ahead with it ?

**Comment by alexhallam** [comment]
@RishabhGarg108 please do.

**Comment by gauravghati** [comment]
Hey @alexhallam, can I work on STL decomposition?

**Comment by gauravghati** [comment]
Hey @alexhallam @rcurtin, can I work on the Seasonal Naive model?

**Comment by alexhallam** [comment]
Yes

**Comment by ashwin2000** [comment]
Hey @rcurtin @alexhallam , have all the methods listed above have either open PRs or been implemented already? If not I'm interested in taking one up.

**Comment by RishabhGarg108** [comment]
Hello @ashwin2000 , please feel free to take any of the methods in the list other than the "naive model". For naive model, there is already a PR #2789 . You can take a look at it if you need any reference.

**Comment by AditiThirdEye** [comment]
Hello @alexhallam, @rcurtin 
If the PR is not done then, I would like to work on this thread on the exponential smoothing model for time series.
Regards,
Aditi

**Comment by AditiThirdEye** [comment]
Hello @alexhallam, @rcurtin
If the PR is not done then, can I work on this thread on the exponential smoothing model for time series?
Regards,
Aditi

**Comment by RishabhGarg108** [comment]
Hey @AditiThirdEye, feel free to work on exponential smoothing.

**Comment by alexhallam** [comment]
Please do.

**Comment by AditiThirdEye** [comment]
Hello @alexhallam , 
Can you please look at the implementation done for the exponential smoothing feature made by me in this [PR](#3046).

**Comment by alexhallam** [comment]
Hi @AditiThirdEye that PR link sends me back to this issue. Can you send another link of tell me the branch name?

**Comment by AditiThirdEye** [comment]
Sorry for that, I changed it now 
https://github.com/mlpack/mlpack/pull/3046
Please take a look, sorry for the inconveniences.

**Comment by cheapkai** [comment]
Hello,

@alexhallam , @rcurtin  could I implement the Holt Winters Forecasting method ?

**Comment by alexhallam** [comment]
Go ahead!

**Comment by rcurtin** [comment]
@cheapkai it would probably be best if you started with a general design of the API before implementation---we'll need to make sure that whatever you implement fits well into the mlpack codebase. :+1:

**Comment by cheapkai** [comment]
Ok! Am starting on it.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by Ris-Bali** [comment]
Hello @rcurtin @alexhallam if there is no open PR , could I implement the ARIMA method for Time Series Forcasting 
Thanks .

**Comment by Ris-Bali** [comment]
Okay I'll work on PR for ARIMA

**Comment by rcurtin** [comment]
@Ris-Bali the most important thing about any PR that adds new support is that the API matches the rest of the mlpack methods.  So please take a look at the rest and be sure that the design you implement is coherent. :+1:

**Comment by Ris-Bali** [comment]
@rcurtin  sorry for disturbing but does the library hai AR(Autoregressive) model  and MA (moving average ) model which are part of ARIMA , I checked in statsnmodels and AR is also used as a forcasting method and it is part of ARIMA ; infact the above list doesn't contain some forcasting methods . So the list above can be extended by adding them . I am working on a PR for AR model first if it has not been implemented .
Thanks

**Comment by alexhallam** [comment]
@Ris-Bali 

> infact the above list doesn't contain some forecasting methods

The list is not exhaustive, new forecasting models come out all the time (see [International Journal of Forecasting](https://www.sciencedirect.com/journal/international-journal-of-forecasting)). The list above is really just to make sure that `mlpack` is not turning a blind eye to forecasting. There is no single package that contains all methods. My preference would be for MLPack to lean more on the side of simple methods. If you are interested in why I think simple methods are important see my pontifications [why-run-simple-methods](https://github.com/alexhallam/tablespoon#why-run-simple-methods) 😆

As far as how ARIMA interacts with AR and MA there are some subtleties despite the acronym. I would suggested reading this https://otexts.com/fpp3/arima-r.html. I would read through most - if not all - of that book if you want to get a sense of how the different models do or do not relate. It is not required, just could be useful context. **MA** for example stands for **Moving Average**, but when used in the context of ARI(MA) the meaning is different. It is not a *simple* moving average of the time series. Instead it uses past residuals. https://otexts.com/fpp3/MA.html.  Things like that are good to be aware of.

**Comment by rcurtin** [comment]
@Ris-Bali note also that there has been some previous work for an exponential smoothing model that was never merged: https://github.com/mlpack/mlpack/pull/3046.  You could, for instance, pick that up too.  (Honestly it is a little disappointing to put a bunch of time into reviewing a PR for it to get so close, but never get merged, so personally I tend to prefer reviewing PRs that build on previous unmerged work and get them merged, instead of entirely new development.)

**Comment by Ris-Bali** [comment]
@alexhallam thank you for those resources; they helped me gain some intuition about the methods; also was not aware simple methods had usecases in the industry ,thanks for that info too !!

**Comment by Ris-Bali** [comment]
> @Ris-Bali note also that there has been some previous work for an exponential smoothing model that was never merged: #3046. You could, for instance, pick that up too. (Honestly it is a little disappointing to put a bunch of time into reviewing a PR for it to get so close, but never get merged, so personally I tend to prefer reviewing PRs that build on previous unmerged work and get them merged, instead of entirely new development.)

Sure I'll try to implement that first.
